File: frontend/src/api/marketData.ts
import { api } from "./client";

export interface Dataset {
  source: string; broker: string; symbol: string; market_type: string; timeframe: string;
}
export interface Coverage extends Dataset { candle_count: number; first_candle: string; last_candle: string }
export interface Page<T> { items: T[]; total: number; offset: number; limit: number }
export interface Issue { code: string; row?: number; message: string }
export interface ImportReport extends Dataset {
  id: number; import_id: number; status: string; file_name: string; file_sha256: string;
  rows_received: number; rows_valid: number; rows_inserted: number; rows_duplicates: number;
  rows_rejected: number; rows_conflicting: number; created_at: string; errors: Issue[]; warnings: Issue[];
}
export interface Candle extends Dataset {
  id: number; import_id: number | null; open_time: string; close_time: string;
  open: string; high: string; low: string; close: string; tick_volume: string | null; spread: string | null;
}
export interface CatalogFilter extends Dataset { start: string; end: string; pattern_length: number; include_doji: boolean }
export interface Pattern {
  pattern: string; pattern_length: number; sample_size: number;
  next_bullish_count: number; next_bearish_count: number; next_doji_count: number;
  next_bullish_probability: string; next_bearish_probability: string; next_doji_probability: string;
  first_observation: string; last_observation: string; distinct_days: number;
}
export interface CatalogResult {
  metadata: CatalogFilter; candles_examined: number; eligible_windows: number;
  windows_skipped_due_to_gaps: number; windows_skipped_due_to_doji: number;
  patterns: Pattern[]; limitation: string;
}
export function datasetOnly(d: Dataset): Dataset {
  return { source: d.source, broker: d.broker, symbol: d.symbol, market_type: d.market_type, timeframe: d.timeframe };
}
export function params(values: object) {
  return new URLSearchParams(Object.entries(values).map(([k, v]) => [k, String(v)])).toString();
}
export const marketApi = {
  options: () => api<{ timeframes: string[]; market_types: string[]; max_upload_mb: number; max_rows: number }>("/market-data/options"),
  coverage: (offset = 0) => api<Page<Coverage>>("/market-data/coverage?" + params({ offset, limit: 100 })),
  imports: (offset = 0) => api<Page<ImportReport>>("/market-data/imports?" + params({ offset, limit: 20 })),
  upload: (body: FormData) => api<ImportReport>("/market-data/imports/csv", { method: "POST", body }),
  candles: (dataset: Coverage, offset = 0) => api<Page<Candle>>("/market-data/candles?" + params({
    ...datasetOnly(dataset), start: dataset.first_candle,
    end: new Date(new Date(dataset.last_candle).getTime() + 1).toISOString(), offset, limit: 20
  })),
  analyze: (filters: CatalogFilter) => api<CatalogResult>("/cataloger/patterns?" + params(filters)),
};

File: backend/app/api/market_data.py
from datetime import datetime
from decimal import Decimal
from fastapi import APIRouter, Depends, File, Form, UploadFile, HTTPException, Query
from fastapi.encoders import jsonable_encoder
from pydantic import ValidationError
from sqlalchemy import select, func
from .deps import db, current_user, require_admin
from ..config import settings
from ..models import Candle, MarketDataImport
from ..market_data.normalization import Dataset, stored_utc
from ..market_data.timeframe import TIMEFRAMES
from ..market_data.ingestion import ingest
from ..market_data.repository import coverage, candle_query, candle_data
from ..cataloger.schemas import CandleQuery, CatalogQuery
from ..cataloger.engine import analyze

router = APIRouter(tags=["market-data"])


def encoded(value):
    return jsonable_encoder(value, custom_encoder={Decimal: str, datetime: lambda d: stored_utc(d).isoformat()})


def import_out(batch):
    values = {column.name: getattr(batch, column.name) for column in MarketDataImport.__table__.columns}
    values["import_id"] = batch.id
    values["errors"] = (batch.error_summary or {}).get("errors", [])
    values["warnings"] = (batch.error_summary or {}).get("warnings", [])
    return values


@router.get("/market-data/options")
def options(user=Depends(current_user)):
    return {
        "timeframes": list(TIMEFRAMES),
        "market_types": ["REGULAR", "OTC"],
        "max_upload_mb": settings.market_data_max_upload_mb,
        "max_rows": settings.market_data_max_rows,
    }


@router.post("/market-data/imports/csv")
def import_csv(
    file: UploadFile = File(...),
    source: str = Form(...),
    symbol: str = Form(...),
    market_type: str = Form(...),
    timeframe: str = Form(...),
    broker: str = Form("UNSPECIFIED"),
    user=Depends(require_admin),
    session=Depends(db),
):
    try:
        dataset = Dataset(source=source, broker=broker, symbol=symbol, market_type=market_type, timeframe=timeframe)
    except ValidationError as exc:
        raise HTTPException(422, str(exc))
    try:
        raw = file.file.read(settings.market_data_max_upload_mb * 1024 * 1024 + 1)
    finally:
        file.file.close()
    if len(raw) > settings.market_data_max_upload_mb * 1024 * 1024:
        raise HTTPException(413, "CSV exceeds configured upload limit")
    # Content is validated independently of filename and client MIME type.
    batch = ingest(session, user.id, dataset, raw, file.filename, settings.market_data_max_rows)
    return encoded(import_out(batch) | {"coverage": coverage(session, dataset)})


@router.get("/market-data/imports")
def imports(limit: int = Query(20, ge=1, le=100), offset: int = Query(0, ge=0), user=Depends(current_user), session=Depends(db)):
    rows = session.scalars(
        select(MarketDataImport).order_by(MarketDataImport.created_at.desc(), MarketDataImport.id.desc()).offset(offset).limit(limit)
    )
    return encoded(
        {
            "items": [import_out(row) for row in rows],
            "total": session.scalar(select(func.count()).select_from(MarketDataImport)),
            "limit": limit,
            "offset": offset,
        }
    )


@router.get("/market-data/coverage")
def dataset_coverage(limit: int = Query(100, ge=1, le=500), offset: int = Query(0, ge=0), user=Depends(current_user), session=Depends(db)):
    groups = select(*[getattr(Candle, k) for k in Dataset.model_fields]).distinct().subquery()
    return encoded(
        {
            "items": coverage(session, offset=offset, limit=limit),
            "total": session.scalar(select(func.count()).select_from(groups)),
            "limit": limit,
            "offset": offset,
        }
    )


@router.get("/market-data/candles")
def candles(query: CandleQuery = Query(), user=Depends(current_user), session=Depends(db)):
    statement = candle_query(query.dataset(), query.start, query.end)
    rows = list(session.scalars(statement.offset(query.offset).limit(query.limit)))
    total = session.scalar(select(func.count()).select_from(statement.order_by(None).subquery()))
    return encoded(
        {
            "items": [dict(id=c.id, import_id=c.import_id, **candle_data(c).__dict__) for c in rows],
            "total": total,
            "limit": query.limit,
            "offset": query.offset,
        }
    )


@router.get("/cataloger/patterns")
def patterns(query: CatalogQuery = Query(), user=Depends(current_user), session=Depends(db)):
    statement = candle_query(query.dataset(), query.start, query.end)
    total = session.scalar(select(func.count()).select_from(statement.order_by(None).subquery()))
    if total > settings.market_data_max_catalog_candles:
        raise HTTPException(422, f"Range exceeds {settings.market_data_max_catalog_candles} candles; narrow dates")
    rows = session.scalars(statement.limit(settings.market_data_max_catalog_candles + 1).execution_options(yield_per=1000))
    try:
        result = analyze((candle_data(c) for c in rows), query.pattern_length, query.include_doji)
    except ValueError as exc:
        raise HTTPException(422, f"Dataset quality prevents cataloging: {exc}")
    if result["candles_examined"] > settings.market_data_max_catalog_candles:
        raise HTTPException(422, "Dataset grew during query; narrow dates and retry")
    return encoded(
        {
            "metadata": query.model_dump(exclude={"limit", "offset"}),
            **result,
            "limitation": "Historical descriptive frequencies only. Overlapping windows may be autocorrelated; no causal or profitability claim.",
        }
    )
